Userland/Libraries/LibWeb/HTML/HTMLDialogElement.cpp

```diff
@@ -93,10 +107,30 @@ WebIDL::ExceptionOr<void> HTMLDialogElement::show_modal()
     if (!document().top_layer_elements().contains(*this))
         document().add_an_element_to_the_top_layer(*this);
 
-    // FIXME: 9. Set this's close watcher to the result of establishing a close watcher given this's relevant global object, with:
-    //            - cancelAction being to return the result of firing an event named cancel at this, with the cancelable
-    //              attribute initialized to true.
-    //            - closeAction being to close the dialog given this and null.
+    // 9. Set this's close watcher to the result of establishing a close watcher given this's relevant global object
+    m_close_watcher = CloseWatcher::establish(*document().window());
+    // - cancelAction given canPreventClose being to return the result of firing an event named cancel at this, with the cancelable attribute initialized to canPreventClose.
+    auto cancel_callback_function = JS::NativeFunction::create(
+        realm(), [this](JS::VM& vm) {
+            auto event_value = vm.argument(0);
+            bool can_prevent_close = MUST(event_value.get(vm, "cancelable")).as_bool();
```

Comment:
Hmm, this looks a little rickety. Do we know for sure this can only ever be called with an object that has a `cancelable` property?

Reply:
It's an event listener so it should always be called with an event object afaik?

Reply:
Hmm, all right. Well if that's the case, if that's the case we should be able to `auto& typed_event = verify_cast<YourDesiredEventType>(event_value.as_object());` :)

Reply:
I tried that before and got a compile error. Just realised I did `auto event` not `auto& event` 🤦🏻 thanks!